Forge.TreeWalker/src/TreeWalkerSession.cs

```diff
@@ -321,18 +358,39 @@ public async Task<string> WalkTree(string treeNodeKey)
                     }
                     finally
                     {
-                        // Always call this callback, whether or not visit node was successful.
-                        await this.Parameters.Callbacks.AfterVisitNode(
+                        if (treeNodeContext_BeforeVisitNode != null)
+                        {
+                            // When treeNodeContext is not null, continues with the new ITreeWalkerCallbacksV2.
+                            TreeNodeContext treeNodeContext_AfterVisitNode = new TreeNodeContext(
+                                this.Parameters.SessionId,
+                                current,
+                                await this.EvaluateDynamicProperty(this.Schema.Tree[current].Properties, null),
+                                this.Parameters.UserContext,
+                                this.walkTreeCts.Token,
+                                this.Parameters.TreeName,
+                                this.Parameters.RootSessionId,
+                                currentNodeSkipActionContext: null
+                            );
+
+                            await this.Parameters.CallbacksV2.AfterVisitNode(treeNodeContext_AfterVisitNode).ConfigureAwait(false);
+                        }
+                        else
+                        {
+                            // Always call this callback, whether or not visit node was successful.
+                            await this.Parameters.Callbacks.AfterVisitNode(
                             this.Parameters.SessionId,
                             current,
                             await this.EvaluateDynamicProperty(this.Schema.Tree[current].Properties, null),
                             this.Parameters.UserContext,
                             this.Parameters.TreeName,
                             this.Parameters.RootSessionId,
                             this.walkTreeCts.Token).ConfigureAwait(false);
+                        }
                     }
 
                     current = next;
+                    this.currentNodeSkipActionContext = null;   // Clear the context of SkipAction and get ready for the next tree node.
```

Comment:
>this.currentNodeSkipActionContext = null; [](start = 20, length = 41)

Let's move this inside the finally block. We want to make sure this gets cleared.

Reply:
fixed

---
In reply to: [505830289](https://github.com/microsoft/Forge/pull/32#discussion_r505830289) [](ancestors = 505830289)